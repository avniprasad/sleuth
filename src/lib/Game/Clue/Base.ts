import type { Loadable, Saveable } from "../../Storage";
import type { Board } from "../Board";

export const enum ClueType {
  Exact,
  Same,
  Adjacent,
  Before,
  Sequential,
}

export const SAVE_CLUE_KEY = "SAVE_CLUE_KEY";

export abstract class Clue implements Saveable, Loadable {
  abstract type: ClueType;

  constructor(public tiles: [number, number][]) {
    this.tiles = tiles;
  }

  abstract apply(board: Board): void;
  abstract validate(board: Board): boolean;

  get key() {
    return SAVE_CLUE_KEY;
  }

  toStorageString(): string {
    return `${this.type}|${this.tiles.map((t) => t.join(":")).join("|")}`;
  }

  fromStorageString(data: string | null): void {
    if (!data) throw Error("empty data when restoring clue");
    const [_, ...tilesS] = data.split("|").filter((d) => d);
    this.tilesFromStorageStrings(tilesS)
  }

  protected tilesFromStorageStrings(data: string[]) {
    this.tiles = data.map((data) => {
      return data
        .split(":")
        .filter((d) => d)
        .map((d) => parseInt(d)) as [number, number];
    });
  }
}
