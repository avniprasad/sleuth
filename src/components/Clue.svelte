<script lang="ts" context="module">
  import { ClueType, type Clue } from "../lib/Game/Clue";
  const TypeSymbols: Record<ClueType, string> = {
    [ClueType.Same]: "|",
    [ClueType.Adjacent]: "-",
    [ClueType.Exact]: "!",
    [ClueType.Sequential]: "=",
    [ClueType.Before]: "<",
  };
</script>

<script lang="ts">
  import clsx from "clsx";

  import type { RowDef } from "../app";
  import Base from "./Cell/Base.svelte";

  export let clue: Clue;
  export let rowDef: RowDef;
  export let finished = false;

  $: tile1 = clue.tiles[0];
  $: tile1Def = rowDef[tile1[0]];
  $: tile2 = clue.tiles[1] || [];
  $: tile2Def = rowDef[tile2[0]] || [];

  console.log(!$clue)
</script>

<div class={clsx("grid grid-cols-3 clue border-2", finished && !$clue && "border-red-500")}>
  <Base square visible class={tile1Def[1]}>
    {tile1Def[0][tile1[1]]}
  </Base>
  {#if clue.type !== ClueType.Sequential}
    <Base square visible class="fill-white">{TypeSymbols[clue.type]}</Base>
  {/if}
  <Base square visible class={tile2Def[1]}>
    {tile2Def[0][tile2[1]]}
  </Base>
  {#if clue.type === ClueType.Sequential}
    {@const tile = clue.tiles[2]}
    {@const def = rowDef[tile[0]]}
    <Base square visible class={def[1]}>
      {def[0][tile[1]]}
    </Base>
  {/if}
</div>
