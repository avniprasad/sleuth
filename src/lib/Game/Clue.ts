import { remove_from_board, type iBoard, type iSolvedBoard } from "./Board";
import { shuffle, cell_is } from "../utils";

export const enum ClueType {
  Exact,
  Same,
  Adjacent,
  Before,
  Sequential,
}

interface BaseClue {
  type: ClueType;
  tiles: [number, number][];
}

interface ExactClue extends BaseClue {
  type: ClueType.Exact;
  tiles: [[number, number]];
  col: number;
}

function applyExact(clue: ExactClue, board: iBoard): boolean {
  const [[row, id]] = clue.tiles;
  const col = clue.col;
  return remove_from_board(board, row, col, id, true);
}

interface AdjacentClue extends BaseClue {
  type: ClueType.Adjacent;
  tiles: [[number, number], [number, number]];
}

function applyAdjacent(clue: AdjacentClue, board: iBoard) { }

interface SameClue extends BaseClue {
  type: ClueType.Same;
  tiles: [[number, number], [number, number]];
}

function applySame(clue: SameClue, board: iBoard) {
  let changed = false;
  const [[row1, id1], [row2, id2]] = clue.tiles;
  for (let i = 0; i < board.length; i++) {
    const cell1 = board[row1][i];
    const cell2 = board[row2][i];
    // may not be needed
    // if (cell_is(cell1, id1) && cell_is(cell2, id2)) {
    //   return false;
    // }
    if (cell_is(cell1, id1)) {
      return remove_from_board(board, row2, i, id2, true);
    }
    if (cell_is(cell2, id2)) {
      return remove_from_board(board, row1, i, id1, true);
    }
    if (!cell1.has(id1)) {
      changed = remove_from_board(board, row2, i, id2);
    }
    if (!cell2.has(id2)) {
      changed = remove_from_board(board, row1, i, id1);
    }
  }
}

interface BeforeClue extends BaseClue {
  type: ClueType.Before;
  tiles: [[number, number], [number, number]];
}

function applyBefore(clue: BeforeClue, board: iBoard) { }

interface SequentialClue extends BaseClue {
  type: ClueType.Sequential;
  tiles: [[number, number], [number, number], [number, number]];
}

function applySequential(clue: SequentialClue, board: iBoard) { }

export type Clue =
  | ExactClue
  | AdjacentClue
  | SameClue
  | BeforeClue
  | SequentialClue;

// Generate a list of all possible clues for the given solved board.
export function generate_clues(board: iSolvedBoard): Clue[] {
  const clues: Clue[] = [];
  for (let r = 0; r < board.length; r++) {
    for (let c = 0; c < board[r].length; c++) {
      const id = board[r][c];
      // Exact clues.
      clues.push({
        type: ClueType.Exact,
        tiles: [[r, id]],
        col: c,
      });
      // Same clues.
      for (let r2 = r + 1; r2 < board.length; r2++) {
        clues.push({
          type: ClueType.Same,
          tiles: [
            [r, id],
            [r2, board[r2][c]],
          ],
        });
      }
      if (c + 1 < board[r].length) {
        // Adjacent clues.
        for (let r2 = 0; r2 < board.length; r2++) {
          clues.push({
            type: ClueType.Adjacent,
            tiles: swap([
              [r, id],
              [r2, board[r2][c + 1]],
            ]),
          });
          if (c > 0) {
            // Sequential clues.
            clues.push({
              type: ClueType.Sequential,
              tiles: swap([
                [r2, board[r2][c - 1]],
                [r, id],
                [r2, board[r2][c + 1]],
              ]),
            });
          }
        }
      }
      // Before clues.
      for (let r2 = 0; r2 < board.length; r2++) {
        for (let c2 = c + 1; c2 < board[r2].length; c2++) {
          clues.push({
            type: ClueType.Before,
            tiles: [
              [r, id],
              [r2, board[r2][c2]],
            ],
          });
        }
      }
    }
  }
  return shuffle(clues);
}

function swap<A extends [unknown, unknown, ...unknown[]]>(a: A): A {
  if (Math.random() < 0.5) {
    const x = a[a.length - 1];
    a[a.length - 1] = a[0];
    a[0] = x;
  }
  return a;
}

export function apply(clue: Clue, board: iBoard): boolean {
  switch (clue.type) {
    case ClueType.Exact: {
      return applyExact(clue, board);
    }
    case ClueType.Sequential: {
      return applySequential(clue, board);
    }
    case ClueType.Same: {
      return applySame(clue, board);
    }
    case ClueType.Before: {
      return applyBefore(clue, board);
    }
    case ClueType.Adjacent: {
      return applyAdjacent(clue, board);
    }
  }
}
