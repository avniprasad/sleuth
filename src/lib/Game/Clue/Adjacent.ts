import type { Board } from "../Board";
import { Clue, ClueType } from "./Base";

export class AdjacentClue extends Clue {
  type: ClueType = ClueType.Adjacent;

  constructor(public tiles: [[number, number], [number, number]]) {
    super(tiles);
  }

  apply(board: Board): void {
    for (let c = 0; c < board.length; c++) {
      for (let [one, two] of [
        [0, 1],
        [1, 0],
      ]) {
        const [row1, id1] = this.tiles[one];
        const [row2, id2] = this.tiles[two];
        const cell = board.get(row1, c);
        const cellL = board.get(row2, c - 1);
        const cellR = board.get(row2, c + 1);
        if (cell?.is(id1)) {
          board.remove(row2, c, id2);
          if (!cellL?.has(id2)) {
            board.remove(row2, c + 1, id2, true);
            return;
          }
          if (!cellR?.has(id2)) {
            board.remove(row2, c - 1, id2, true);
            return;
          }
        } else if (!cellL?.has(id2) && !cellR?.has(id2)) {
          board.remove(row1, c, id1);
        }
      }
    }
  }

  validate(board: Board): boolean {
    const [[row1, id1], [row2, id2]] = this.tiles;
    for (let i = 1; i < board.length - 1; i++) {
      if (board.get(row1, i)?.is(id1)) {
        return (
          !!board.get(row2, i - 1)?.is(id2) || !!board.get(row2, i + 1)?.is(id2)
        );
      }
      if (board.get(row2, i)?.is(id2)) {
        return (
          !!board.get(row1, i - 1)?.is(id1) || !!board.get(row1, i + 1)?.is(id1)
        );
      }
    }
    return false;
  }
}
