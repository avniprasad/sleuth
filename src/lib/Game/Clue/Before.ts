import type { Board } from "../Board";
import { Clue, ClueType } from "./Base";

export class BeforeClue extends Clue {
  type: ClueType = ClueType.Before;

  constructor(public tiles: [[number, number], [number, number]]) {
    super(tiles);
  }

  apply(board: Board): void {
    const [[row1, id1], [row2, id2]] = this.tiles;
    let found_fist = false;
    let found_second = false;
    let c = 0;
    for (; c < board.length; c++) {
      const cell1 = board.get(row1, c);
      if (!found_fist) {
        board.remove(row2, c, id2);
      }
      if (cell1?.has(id1)) {
        found_fist = true;
        break;
      }
    }
    for (let c2 = board.length - 1; c2 > c; c2--) {
      const cell2 = board.get(row2, c2);
      if (!found_second) {
        board.remove(row1, c2, id1);
      }
      if (cell2?.has(id2)) {
        break;
      }
    }
  }

  validate(board: Board): boolean {
    const [[row1, id1], [row2, id2]] = this.tiles;
    let found_first = false
    for (let i = 0; i < board.length; i++) {
      found_first ||= !!board.get(row1, i)?.is(id1)
      if (board.get(row2, i)?.is(id2)) {
        return found_first;
      }
    }
    return false;
  }
}
