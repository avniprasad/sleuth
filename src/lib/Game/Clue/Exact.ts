import type { Board } from "../Board";
import { Clue, ClueType } from "./Base";

export class ExactClue extends Clue {
  type: ClueType = ClueType.Exact;

  constructor(
    public tiles: [[number, number]],
    public col: number,
  ) {
    super(tiles);
  }

  apply(board: Board): void {
    const [[row, id]] = this.tiles;
    const col = this.col;
    board.remove(row, col, id, true);
  }

  validate(board: Board): boolean {
    const [[row, id]] = this.tiles;
    return !!board.get(row, this.col)?.is(id);
  }

  toStorageString(): string {
    return `${this.type}|${this.col}|${this.tiles
      .map((t) => t.join(":"))
      .join("|")}`;
  }

  fromStorageString(data: string | null): void {
    if (!data) throw Error("empty data when restoring clue");
    const [_, col, ...tilesS] = data.split("|").filter((d) => d);
    this.col = parseInt(col);
    this.tilesFromStorageStrings(tilesS);
  }
}
