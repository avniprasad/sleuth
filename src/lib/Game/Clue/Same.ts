import type { Board } from "../Board";
import { Clue, ClueType } from "./Base";

export class SameClue extends Clue {
  type: ClueType = ClueType.Same;

  constructor(public tiles: [[number, number], [number, number]]) {
    super(tiles);
  }

  apply(board: Board): void {
    const [[row1, id1], [row2, id2]] = this.tiles;
    for (let c = 0; c < board.length; c++) {
      const cell1 = board.get(row1, c);
      const cell2 = board.get(row2, c);
      if (cell1?.is(id1)) {
        board.remove(row2, c, id2, true);
        return;
      }
      if (cell2?.is(id2)) {
        board.remove(row1, c, id1, true);
        return;
      }
      if (!cell1?.has(id1)) {
        board.remove(row2, c, id2);
      }
      if (!cell2?.has(id2)) {
        board.remove(row1, c, id1);
      }
    }
  }

  _validate(board: Board): boolean {
    const [[row1, id1], [row2, id2]] = this.tiles;
    for (let i = 0; i < board.length; i++) {
      if (board.get(row1, i)?.is(id1)) {
        return !!board.get(row2, i)?.is(id2);
      }
      if (board.get(row2, i)?.is(id2)) {
        return !!board.get(row1, i)?.is(id1);
      }
    }
    return false;
  }
}
