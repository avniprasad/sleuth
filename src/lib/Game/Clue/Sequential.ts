import type { Board } from "../Board";
import { Clue, ClueType } from "./Base";

export class SequentialClue extends Clue {
  type: ClueType = ClueType.Sequential;

  constructor(
    public tiles: [[number, number], [number, number], [number, number]],
  ) {
    super(tiles);
  }

  apply(board: Board): void {
    for (let c = 0; c < board.length; c++) {
      const [row1, id1] = this.tiles[0];
      const [rowMid, idMid] = this.tiles[1];
      const [row2, id2] = this.tiles[2];
      const cellMid = board.get(rowMid, c);
      if (cellMid?.is(idMid)) {
        board.remove(row1, c, id1);
        board.remove(row2, c, id2);
      } else if (
        (!board.get(row1, c - 1)?.has(id1) &&
          !board.get(row2, c - 1)?.has(id2)) ||
        (!board.get(row1, c + 1)?.has(id1) && !board.get(row2, c + 1)?.has(id2))
      ) {
        board.remove(rowMid, c, idMid);
      }
      for (let [me, other] of [
        [0, 2],
        [2, 0],
      ]) {
        const [row1, id1] = this.tiles[me];
        const [row2, id2] = this.tiles[other];
        const cell1 = board.get(row1, c - 1);
        const cell2 = board.get(row2, c + 1);
        if (cellMid?.is(idMid)) {
          if (!cell1?.has(id1) || !cell2?.has(id2)) {
            board.remove(row2, c - 1, id2, true);
            board.remove(row1, c + 1, id1, true);
            return;
          }
        } else if (cell1?.is(id1) && cell2?.is(id2)) {
          board.remove(rowMid, c, idMid, true);
          return;
        }
        const cell = board.get(row1, c);
        const cellMidR = board.get(rowMid, c + 1);
        const cellMidL = board.get(rowMid, c - 1);
        if (cell?.is(id1)) {
          if (!cellMidL?.has(idMid)) {
            board.remove(rowMid, c + 1, idMid, true);
            board.remove(row2, c + 2, id2, true);
            return;
          }
          if (!cellMidR?.has(idMid)) {
            board.remove(rowMid, c - 1, idMid, true);
            board.remove(row2, c - 2, id2, true);
            return;
          }
        }
      }
    }
  }

  _validate(board: Board): boolean {
    const [[row1, id1], [rowMid, idMid], [row2, id2]] = this.tiles;
    for (let i = 0; i < board.length; i++) {
      if (board.get(rowMid, i)?.is(idMid)) {
        return !!(
          (board.get(row1, i - 1)?.is(id1) &&
            board.get(row2, i + 1)?.is(id2)) ||
          (board.get(row1, i + 1)?.is(id1) && board.get(row2, i - 1)?.is(id2))
        );
      }
    }
    return false;
  }
}
