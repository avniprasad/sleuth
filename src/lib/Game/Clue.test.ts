import { describe, test, expect } from "vitest";
import { ClueType, generate_clues } from "./Clue";

describe("generate_clues", () => {
  test("generates all clues", () => {
    const board = [
      [0, 1, 2],
      [0, 1, 2],
      [0, 1, 2],
    ];
    const rand = Math.random;
    Math.random = () => 0.5;

    const clues = generate_clues(board);

    const exact = clues.filter(({ type }) => type === ClueType.Exact);
    expect(exact.length).toBe(9);
    expect(exact).toMatchSnapshot("exact clues");

    const same = clues.filter(({ type }) => type === ClueType.Same);
    expect(same.length).toBe(9);
    expect(same).toMatchSnapshot("same clues");

    const adj = clues.filter(({ type }) => type === ClueType.Adjacent);
    expect(adj.length).toBe(18);
    expect(adj).toMatchSnapshot("adj clues");

    const before = clues.filter(({ type }) => type === ClueType.Before);
    expect(before.length).toBe(27);
    expect(before).toMatchSnapshot("before clues");

    const sequential = clues.filter(({ type }) => type === ClueType.Sequential);
    expect(sequential.length).toBe(27);
    expect(sequential).toMatchSnapshot("sequential clues");

    Math.random = rand;
  });
});

describe("apply_clue", () => {
  describe("Exact", () => {
    test("works");
  });
  describe("Same", () => {
    test("works");
  });
  describe("Adjacent", () => {
    test("works");
  });
  describe("Sequential", () => {
    test("works");
  });
  describe("Before", () => {
    test("works");
  });
});
