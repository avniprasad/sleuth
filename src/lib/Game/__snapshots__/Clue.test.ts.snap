// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate_clues > generates all clues > adj clues 1`] = `
[
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        0,
        1,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        1,
        1,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        2,
        1,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        0,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        0,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        0,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        0,
        1,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        1,
        1,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        2,
        1,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        1,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        1,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        1,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        0,
        1,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        1,
        1,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        2,
        1,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        2,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        2,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 2,
  },
  {
    "tiles": [
      [
        2,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 2,
  },
]
`;

exports[`generate_clues > generates all clues > before clues 1`] = `
[
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        0,
        1,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        1,
        1,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        2,
        1,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        0,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        0,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        0,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        0,
        1,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        1,
        1,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        2,
        1,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        1,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        1,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        1,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        0,
        1,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        1,
        1,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        2,
        1,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        2,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        2,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 3,
  },
  {
    "tiles": [
      [
        2,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 3,
  },
]
`;

exports[`generate_clues > generates all clues > exact clues 1`] = `
[
  {
    "col": 0,
    "tiles": [
      [
        0,
        0,
      ],
    ],
    "type": 0,
  },
  {
    "col": 1,
    "tiles": [
      [
        0,
        1,
      ],
    ],
    "type": 0,
  },
  {
    "col": 2,
    "tiles": [
      [
        0,
        2,
      ],
    ],
    "type": 0,
  },
  {
    "col": 0,
    "tiles": [
      [
        1,
        0,
      ],
    ],
    "type": 0,
  },
  {
    "col": 1,
    "tiles": [
      [
        1,
        1,
      ],
    ],
    "type": 0,
  },
  {
    "col": 2,
    "tiles": [
      [
        1,
        2,
      ],
    ],
    "type": 0,
  },
  {
    "col": 0,
    "tiles": [
      [
        2,
        0,
      ],
    ],
    "type": 0,
  },
  {
    "col": 1,
    "tiles": [
      [
        2,
        1,
      ],
    ],
    "type": 0,
  },
  {
    "col": 2,
    "tiles": [
      [
        2,
        2,
      ],
    ],
    "type": 0,
  },
]
`;

exports[`generate_clues > generates all clues > same clues 1`] = `
[
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        1,
        0,
      ],
    ],
    "type": 1,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        2,
        0,
      ],
    ],
    "type": 1,
  },
  {
    "tiles": [
      [
        0,
        1,
      ],
      [
        1,
        1,
      ],
    ],
    "type": 1,
  },
  {
    "tiles": [
      [
        0,
        1,
      ],
      [
        2,
        1,
      ],
    ],
    "type": 1,
  },
  {
    "tiles": [
      [
        0,
        2,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 1,
  },
  {
    "tiles": [
      [
        0,
        2,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 1,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        2,
        0,
      ],
    ],
    "type": 1,
  },
  {
    "tiles": [
      [
        1,
        1,
      ],
      [
        2,
        1,
      ],
    ],
    "type": 1,
  },
  {
    "tiles": [
      [
        1,
        2,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 1,
  },
]
`;

exports[`generate_clues > generates all clues > sequential clues 1`] = `
[
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        0,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        0,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        0,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        0,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        0,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        0,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        0,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        0,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        0,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        1,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        1,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        1,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        1,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        1,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        1,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        1,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        1,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        1,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        2,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        2,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        0,
        0,
      ],
      [
        2,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        2,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        2,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        1,
        0,
      ],
      [
        2,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        2,
        1,
      ],
      [
        0,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        2,
        1,
      ],
      [
        1,
        2,
      ],
    ],
    "type": 4,
  },
  {
    "tiles": [
      [
        2,
        0,
      ],
      [
        2,
        1,
      ],
      [
        2,
        2,
      ],
    ],
    "type": 4,
  },
]
`;
